engine/src/conversion/analysis/fun/implicit_constructors.rs

```diff
@@ -59,261 +77,481 @@ struct ExplicitFound {
+                            SpecialMemberFound::NotPresent
+                        }
+                        Some(ExplicitFound::UserDefined(visibility)) => {
+                            SpecialMemberFound::Explicit(*visibility)
+                        }
+                    }
+                };
+                let items_found = ItemsFound {
+                    default_constructor: is_explicit(ExplicitKind::DefaultConstructor),
+                    destructor: match find_explicit(ExplicitKind::Destructor) {
+                        // Assume that unknown types have destructors. This is common, and allows
+                        // use to generate UniquePtr wrappers with them.
+                        //
+                        // However, this will generate C++ code that doesn't compile if the unknown
+                        // type does not have an accessible destructor. Maybe we should have a way
+                        // to disable that?
+                        //
+                        // TODO: For https://github.com/google/autocxx/issues/815, map
+                        // ExplicitFound::Defaulted(_) to Explicit.
+                        None => SpecialMemberFound::Implicit,
+                        // If there are multiple destructors, assume that one of them will be
+                        // selected by overload resolution.
+                        Some(ExplicitFound::Multiple) => {
+                            SpecialMemberFound::Explicit(CppVisibility::Public)
+                        }
+                        Some(ExplicitFound::Deleted) => SpecialMemberFound::NotPresent,
+                        Some(ExplicitFound::UserDefined(visibility)) => {
+                            SpecialMemberFound::Explicit(*visibility)
+                        }
+                    },
+                    const_copy_constructor: is_explicit(ExplicitKind::ConstCopyConstructor),
+                    non_const_copy_constructor: is_explicit(ExplicitKind::NonConstCopyConstructor),
+                    move_constructor: is_explicit(ExplicitKind::MoveConstructor),
+                };
+                log::info!(
+                    "Special member functions found for {:?}: {:?}",
+                    name,
+                    items_found
+                );
+                assert!(
+                    all_items_found.insert(name.clone(), items_found).is_none(),
+                    "Duplicate struct: {:?}",
+                    name
+                );
+            } else {
+                // If no user-declared constructors of any kind are provided for a class type (struct, class, or union),
+                // the compiler will always declare a default constructor as an inline public member of its class.
+                //
+                // The implicitly-declared or defaulted default constructor for class T is defined as deleted if any of the following is true:
```

Comment:
I think this no longer meets the original spirit of the split between `implicit_constructors.rs` and `implicit_constructor_rules.rs`. The original idea was that all these C++ rules are encoded within `implicit_constructor_rules.rs`, _independent of any other part of the autocxx implementation_ so that they could have standalone unit tests (which you've deleted). Then `implicit_constructors.rs` was the glue that connected it into autocxx.

I'm totally fine to lose that distinction - I think, at this point, you may as well just merge the files? It's up to you, but I don't think the distinction is worthwhile any more now that most of the rules are encoded within `implicit_constructors.rs` .

Reply:
I was wondering about that. I see how it made sense, but I think at this point it's too tied into known_types and the rest that separate tests are more brittle and prone to not filling everything in correctly than they're worth. Done